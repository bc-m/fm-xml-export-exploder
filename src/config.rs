#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CatalogType {
    Accounts,
    BaseDirectory,
    BaseTable,
    CalcsForCustomFunctions,
    CustomFunctions,
    CustomMenu,
    CustomMenuSet,
    ExternalDataSource,
    ExtendedPrivileges,
    FieldsForTables,
    FileAccess,
    Layout,
    Library,
    OptionsForValueLists,
    PrivilegeSets,
    Relationship,
    Script,
    StepsForScripts,
    TableOccurrence,
    Theme,
    ValueList,
}

impl CatalogType {
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        match bytes {
            b"AccountsCatalog" => Some(Self::Accounts),
            b"BaseDirectoryCatalog" => Some(Self::BaseDirectory),
            b"BaseTableCatalog" => Some(Self::BaseTable),
            b"CalcsForCustomFunctions" => Some(Self::CalcsForCustomFunctions),
            b"CustomFunctionsCatalog" => Some(Self::CustomFunctions),
            b"CustomMenuCatalog" => Some(Self::CustomMenu),
            b"CustomMenuSetCatalog" => Some(Self::CustomMenuSet),
            b"ExternalDataSourceCatalog" => Some(Self::ExternalDataSource),
            b"ExtendedPrivilegesCatalog" => Some(Self::ExtendedPrivileges),
            b"FieldsForTables" => Some(Self::FieldsForTables),
            b"FileAccessCatalog" => Some(Self::FileAccess),
            b"LayoutCatalog" => Some(Self::Layout),
            b"LibraryCatalog" => Some(Self::Library),
            b"OptionsForValueLists" => Some(Self::OptionsForValueLists),
            b"PrivilegeSetsCatalog" => Some(Self::PrivilegeSets),
            b"RelationshipCatalog" => Some(Self::Relationship),
            b"ScriptCatalog" => Some(Self::Script),
            b"StepsForScripts" => Some(Self::StepsForScripts),
            b"TableOccurrenceCatalog" => Some(Self::TableOccurrence),
            b"ThemeCatalog" => Some(Self::Theme),
            b"ValueListCatalog" => Some(Self::ValueList),
            _ => None,
        }
    }

    pub fn get_config(&self) -> CatalogConfig {
        match self {
            Self::Accounts => CatalogConfig {
                catalog_item_name: b"Account".to_vec(),
                out_folder_name: "accounts".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "Account/AccountReference/@id".to_string(),
            },
            Self::BaseDirectory => CatalogConfig {
                catalog_item_name: b"BaseDirectory".to_vec(),
                out_folder_name: "base_directory".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::BaseTable => CatalogConfig {
                catalog_item_name: b"BaseTable".to_vec(),
                out_folder_name: "table_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CalcsForCustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunctionCalc".to_vec(),
                out_folder_name: "custom_function_calcs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "CustomFunctionCalc/CustomFunctionReference/@id".to_string(),
            },
            Self::CustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunction".to_vec(),
                out_folder_name: "custom_function_stubs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::CustomMenu => CatalogConfig {
                catalog_item_name: b"CustomMenu".to_vec(),
                out_folder_name: "custom_menus".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CustomMenuSet => CatalogConfig {
                catalog_item_name: b"CustomMenuSet".to_vec(),
                out_folder_name: "custom_menu_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExternalDataSource => CatalogConfig {
                catalog_item_name: b"ExternalDataSource".to_vec(),
                out_folder_name: "external_data_sources".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExtendedPrivileges => CatalogConfig {
                catalog_item_name: b"ExtendedPrivilege".to_vec(),
                out_folder_name: "extended_privileges".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::FieldsForTables => CatalogConfig {
                catalog_item_name: b"FieldCatalog".to_vec(),
                out_folder_name: "table_fields".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "FieldCatalog/BaseTableReference/@id".to_string(),
            },
            Self::FileAccess => CatalogConfig {
                catalog_item_name: b"Authorization".to_vec(),
                out_folder_name: "file_access".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Layout => CatalogConfig {
                catalog_item_name: b"Layout".to_vec(),
                out_folder_name: "layouts".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "Layout/LayoutReference/@id".to_string(),
            },
            Self::Library => CatalogConfig {
                catalog_item_name: b"BinaryData".to_vec(),
                out_folder_name: "libraries".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "BinaryData/LibraryReference/@id".to_string(),
            },
            Self::OptionsForValueLists => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_options".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "ValueList/ValueListReference/@id".to_string(),
            },
            Self::PrivilegeSets => CatalogConfig {
                catalog_item_name: b"PrivilegeSet".to_vec(),
                out_folder_name: "privilege_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Relationship => CatalogConfig {
                catalog_item_name: b"Relationship".to_vec(),
                out_folder_name: "relationships".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Script => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::StepsForScripts => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_steps".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "Script/ScriptReference/@id".to_string(),
            },
            Self::TableOccurrence => CatalogConfig {
                catalog_item_name: b"TableOccurrence".to_vec(),
                out_folder_name: "table_occurrences".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Theme => CatalogConfig {
                catalog_item_name: b"Theme".to_vec(),
                out_folder_name: "themes".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ValueList => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
        }
    }
}

#[derive(Debug)]
pub struct CatalogConfig {
    pub catalog_item_name: Vec<u8>,
    pub out_folder_name: String,
    pub wrapped_in_object_list: bool,
    pub uses_folders: bool,
    pub id_path: String,
}

pub struct Flags {
    pub parse_all_lines: bool,
    pub lossless: bool,
    pub verbose: Option<Option<String>>,
}
