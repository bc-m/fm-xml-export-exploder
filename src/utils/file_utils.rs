use anyhow::{anyhow, Error, Result};
use std::path::PathBuf;

/// Validate that a path exists and is a directory
pub fn valid_dir_or_throw(dir_path: &PathBuf) -> Result<(), Error> {
    let metadata = std::fs::metadata(dir_path)
        .map_err(|_| anyhow!("Path '{}' not exists", dir_path.display()))?;

    match metadata.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Path '{}' is not a directory", dir_path.display())),
    }
}

/// Join scope ID and name into a formatted string
pub fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

/// Escape filename by replacing special characters with underscores
pub fn escape_filename(filename: &str) -> String {
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

/// Check if a line should be skipped based on content
pub fn should_skip_line(line: &str) -> bool {
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}
