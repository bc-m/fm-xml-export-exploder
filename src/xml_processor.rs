use std::io::{BufRead, BufReader, Read};
use std::path::{Path, PathBuf};
use std::{fs::File, time::Instant};

use anyhow::{bail, Context, Error, Result};
use encoding_rs_io::DecodeReaderBytes;
use quick_xml::events::{BytesEnd, BytesStart, Event};
use quick_xml::reader::Reader;

use crate::catalog::xml_explode_catalog;
use crate::config::{CatalogType, Flags};
use crate::custom_function_sanitizer::create_sanitized_custom_functions;
use crate::script_sanitizer::create_sanitized_scripts;
use crate::supporting::process_supporting_element;
use crate::utils::attributes::get_attribute;
use crate::utils::xml_utils::{end_element_to_string, start_element_to_string, XmlEventType};
use crate::utils::{build_out_dir_path, delete_output_directory, write_xml_file, FolderStructure};
use crate::utils::{create_dir, push_line_to_skeleton};
use crate::Skeleton;

pub enum TopLevelSection {
    Structure,
    Metadata,
    DdrInfo,
}

pub enum Action {
    Add,
    Modify,
    Replace,
    Delete,
}

pub enum Qualifier {
    SanitizedScripts,
    SanitizedCustomFunctions,
}

/// Context for XML catalog processing
pub struct ProcessingContext<'a, R: Read + BufRead> {
    pub reader: &'a mut Reader<R>,
    pub path_stack: &'a mut Vec<Vec<u8>>,
    pub root_out_dir: PathBuf,
    pub saxml_version: Option<String>,
    pub db_name: Option<String>,
    pub top_level_section: Option<TopLevelSection>,
    pub action: Option<Action>,
    pub catalog_type: Option<CatalogType>,
    pub current_out_dir: PathBuf,
    pub skeleton: &'a mut Skeleton,
    pub flags: &'a Flags,
}

/// Process a single XML file and explode it into individual files
pub fn explode_xml(
    fm_export_file_path: &PathBuf,
    root_out_dir: &Path,
    flags: &Flags,
) -> Result<(), Error> {
    let start = Instant::now();
    let fm_export_file_name = fm_export_file_path.file_name().unwrap().to_str().unwrap();

    // Open XML file
    let file = File::open(fm_export_file_path)
        .with_context(|| format!("Error opening file {}", fm_export_file_path.display(),))?;

    // Initialize variables
    let mut depth = 0;
    let mut cf_folder_structure: Option<FolderStructure> = None;
    let mut script_folder_structure: Option<FolderStructure> = None;
    let mut buf = Vec::new();

    // Instantiate processing context which will be passed around to various functions
    let mut context = ProcessingContext {
        reader: &mut Reader::from_reader(BufReader::new(DecodeReaderBytes::new(file))),
        path_stack: &mut Vec::new(),
        root_out_dir: root_out_dir.to_path_buf(),
        saxml_version: None,
        db_name: None,
        top_level_section: None,
        action: None,
        catalog_type: None,
        current_out_dir: PathBuf::new(),
        skeleton: &mut Skeleton::default(),
        flags,
    };

    // Iterate over XML events
    loop {
        match context.reader.read_event_into(&mut buf) {
            Err(e) => {
                println!("Error in {}: {}", fm_export_file_path.display(), e);
                break;
            }
            Ok(Event::Eof) => break,
            Ok(Event::Start(start_tag)) => {
                context.path_stack.push(start_tag.name().as_ref().to_vec());
                push_start_to_skeleton(
                    &start_tag,
                    context.skeleton,
                    context.path_stack,
                    context.flags,
                );
                match depth {
                    0 => {
                        process_root_element(&mut context, &start_tag)?;
                        create_dir(&context.root_out_dir);
                        delete_output_directory(&context)?;
                    }
                    1 => {
                        let was_xml_element_consumed =
                            process_top_level_section(&mut context, &start_tag)?;
                        if was_xml_element_consumed {
                            continue;
                        }
                    }
                    2 => {
                        context.action = match start_tag.name().as_ref() {
                            b"AddAction" => Some(Action::Add),
                            b"ModifyAction" => Some(Action::Modify),
                            b"ReplaceAction" => Some(Action::Replace),
                            b"DeleteAction" => Some(Action::Delete),
                            _ => None,
                        };
                    }
                    3 => {
                        let is_supported_catalog = process_catalog_elements(
                            &mut context,
                            &start_tag,
                            &mut cf_folder_structure,
                            &mut script_folder_structure,
                        )?;
                        if is_supported_catalog {
                            continue;
                        }
                    }
                    _ => {}
                }
                depth += 1;
            }
            Ok(Event::End(e)) => {
                context.path_stack.pop();
                push_end_to_skeleton(&e, context.skeleton, context.path_stack, context.flags);
                depth -= 1;
            }
            _ => {}
        }
        buf.clear()
    }

    // Write skeleton file if in lossless mode
    if flags.lossless {
        write_skeleton_file(&context)?;
    }

    println!(
        "â†’ {} finished in {} ms.",
        fm_export_file_name,
        start.elapsed().as_millis()
    );

    Ok(())
}

fn process_root_element<R: Read + BufRead>(
    context: &mut ProcessingContext<'_, R>,
    e: &BytesStart,
) -> Result<(), Error> {
    match e.name().as_ref() {
        b"FMDynamicTemplate" | b"FMSaveAsXML" => {
            context.db_name = Some(
                get_attribute(e, "File")
                    .unwrap()
                    .strip_suffix(".fmp12")
                    .unwrap()
                    .to_string(),
            );
            context.saxml_version = Some(get_attribute(e, "version").unwrap().to_string());
        }
        _ => {
            bail!("Unsupported XML-format");
        }
    }
    Ok(())
}

/// Process supporting elements (Metadata, DDR_INFO) that are not part of the main structure
fn process_top_level_section<R: Read + BufRead>(
    context: &mut ProcessingContext<'_, R>,
    start_tag: &BytesStart,
) -> Result<bool, Error> {
    let mut was_xml_element_consumed = false;
    match start_tag.name().as_ref() {
        b"Structure" => {
            context.top_level_section = Some(TopLevelSection::Structure);
        }
        b"Metadata" => {
            context.top_level_section = Some(TopLevelSection::Metadata);
            process_supporting_element(context, start_tag, "metadata")?;
            was_xml_element_consumed = true;
        }
        b"DDR_INFO" => {
            context.top_level_section = Some(TopLevelSection::DdrInfo);
            process_supporting_element(context, start_tag, "ddr_info")?;
            was_xml_element_consumed = true;
        }
        _ => context.top_level_section = None,
    }
    Ok(was_xml_element_consumed)
}

fn process_catalog_elements<R: Read + BufRead>(
    context: &mut ProcessingContext<'_, R>,
    start_tag: &BytesStart,
    cf_folder_structure: &mut Option<FolderStructure>,
    script_folder_structure: &mut Option<FolderStructure>,
) -> Result<bool, Error> {
    // Detect catalog type
    let catalog_type = match CatalogType::from_bytes(start_tag.name().as_ref()) {
        Some(catalog_type) => catalog_type,
        None => {
            return Ok(false);
        }
    };
    context.catalog_type = Some(catalog_type);
    // Handle special cases that need folder structures
    let folder_structure = match catalog_type {
        CatalogType::CalcsForCustomFunctions => cf_folder_structure.as_ref(),
        CatalogType::StepsForScripts => script_folder_structure.as_ref(),
        _ => None,
    };

    let xml_out_dir_path = build_out_dir_path(context, None)?;
    let base_dir_path = context.current_out_dir.clone();
    context.current_out_dir = xml_out_dir_path.clone();

    let folder_structure_result = xml_explode_catalog(context, start_tag, folder_structure)?;

    context.current_out_dir = base_dir_path; // Restore to its original value

    // Update folder structures for catalogs that return them
    if let Some(folder_structure) = folder_structure_result {
        match catalog_type {
            CatalogType::CustomFunctions => *cf_folder_structure = Some(folder_structure),
            CatalogType::Script => *script_folder_structure = Some(folder_structure),
            _ => {}
        }
    }

    // Handle post-processing for StepsForScripts
    if catalog_type == CatalogType::StepsForScripts {
        let sanitized_scripts_dir_path =
            build_out_dir_path(context, Some(Qualifier::SanitizedScripts))?;
        create_sanitized_scripts(
            &xml_out_dir_path,
            &sanitized_scripts_dir_path,
            context.flags,
        );
    }

    if catalog_type == CatalogType::CalcsForCustomFunctions {
        let sanitized_cf_dir_path =
            build_out_dir_path(context, Some(Qualifier::SanitizedCustomFunctions))?;
        create_sanitized_custom_functions(&xml_out_dir_path, &sanitized_cf_dir_path);
    }
    Ok(true) // is_supported_catalog
}

fn write_skeleton_file<R: Read + BufRead>(context: &ProcessingContext<'_, R>) -> Result<(), Error> {
    let skeleton_dir_path = build_out_dir_path(context, None)?;
    let skeleton_file_path = skeleton_dir_path.join("skeleton.xml");
    write_xml_file(
        &skeleton_file_path,
        &context.skeleton.content,
        0,
        context.flags,
    );
    Ok(())
}

fn push_start_to_skeleton(
    e: &BytesStart,
    skeleton: &mut Skeleton,
    path_stack: &[Vec<u8>],
    flags: &Flags,
) {
    if flags.lossless {
        match path_stack.len() {
            1..=3 => {
                push_line_to_skeleton(
                    skeleton,
                    path_stack.len(),
                    1,
                    start_element_to_string(e, flags).as_str(),
                    true,
                    XmlEventType::Start,
                );
            }
            4 => {
                if let Some(last) = path_stack.iter().rev().nth(1) {
                    let parent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&parent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            start_element_to_string(e, flags).as_str(),
                            true,
                            XmlEventType::Start,
                        );
                    }
                }
            }
            5 => {
                if let Some(last) = path_stack.iter().rev().nth(2) {
                    let grandparent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&grandparent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            start_element_to_string(e, flags).as_str(),
                            true,
                            XmlEventType::Start,
                        );
                    }
                }
            }
            _ => {}
        }
    }
}

fn push_end_to_skeleton(
    e: &BytesEnd,
    skeleton: &mut Skeleton,
    path_stack: &[Vec<u8>],
    flags: &Flags,
) {
    if flags.lossless {
        match path_stack.len() {
            0..=2 => {
                push_line_to_skeleton(
                    skeleton,
                    path_stack.len(),
                    1,
                    end_element_to_string(e).as_str(),
                    false,
                    XmlEventType::End,
                );
            }
            3 => {
                if let Some(last) = path_stack.iter().next_back() {
                    let parent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&parent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            end_element_to_string(e).as_str(),
                            false,
                            XmlEventType::End,
                        );
                    }
                }
            }
            4 => {
                if let Some(last) = path_stack.iter().rev().nth(1) {
                    let grandparent = std::str::from_utf8(last).unwrap();
                    if ["AddAction", "ModifyAction"].contains(&grandparent) {
                        push_line_to_skeleton(
                            skeleton,
                            path_stack.len(),
                            1,
                            end_element_to_string(e).as_str(),
                            false,
                            XmlEventType::End,
                        );
                    }
                }
            }
            _ => {}
        }
    }
}
