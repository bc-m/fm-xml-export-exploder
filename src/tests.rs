#[cfg(test)]
mod tests {
    use std::fs;
    use std::io::Read;
    use std::path::{Path, PathBuf};

    use insta::assert_snapshot;
    use similar_asserts::assert_eq;
    use walkdir::WalkDir;

    use crate::config::Flags;
    use crate::utils::file_utils::escape_filename;
    use crate::xml_processor::explode_xml;
    use crate::OutputTree;

    #[test]
    fn test_escape_filename() {
        assert_eq!(escape_filename("filename.xml"), "filename.xml");
        assert_eq!(escape_filename("file/name.xml"), "file_name.xml");
        assert_eq!(escape_filename("file|name.xml"), "file_name.xml");
    }

    #[test]
    fn snapshot_test_lossless() {
        snapshot_test_with_mode(true, OutputTree::Db, "snapshots_db_lossless");
    }

    #[test]
    fn snapshot_test_lossy_domain() {
        snapshot_test_with_mode(false, OutputTree::Domain, "snapshots_domain_lossy");
    }

    fn snapshot_test_with_mode(
        is_lossless: bool,
        output_tree: OutputTree,
        snapshot_folder_name: &str,
    ) {
        let path_str = format!("./tests/{}", snapshot_folder_name);
        let snapshot_dir = Path::new(&path_str);
        let input_dir = Path::new("./tests/xml");
        // Make output_dir mode-specific to avoid interference from prior or parallel test runs
        let output_tree_str = match output_tree {
            OutputTree::Db => "db",
            OutputTree::Domain => "domain",
        };
        let path_str = if is_lossless {
            format!("./tests/out_{output_tree_str}_lossless")
        } else {
            format!("./tests/out_{output_tree_str}_lossy")
        };
        let output_dir = Path::new(&path_str);
        let flags = Flags {
            parse_all_lines: false,
            lossless: is_lossless,
            output_tree,
        };
        let _ = fs::remove_dir_all(output_dir);

        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path(snapshot_dir);
        settings.set_prepend_module_to_snapshot(false);

        let paths = fs::read_dir(input_dir)
            .unwrap()
            .filter_map(|entry| entry.ok().map(|e| e.path()))
            .filter(|path| path.is_file())
            .filter(|e| e.file_name().unwrap() != ".DS_Store")
            .collect::<Vec<_>>();

        for path in paths {
            explode_xml(&path, output_dir, &flags)
                .unwrap_or_else(|_| panic!("Error processing file '{}'", path.display()));
        }

        let output_files: Vec<PathBuf> = WalkDir::new(output_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap() != ".DS_Store")
            .map(|e| e.path().to_path_buf())
            .collect();

        for output_file in &output_files {
            let output_content = String::from_utf8(read_file(output_file)).unwrap();
            let output_file_slug = output_file.strip_prefix(output_dir).unwrap();
            let path_str = format!("../tests/{}", snapshot_folder_name);
            let snapshot_path = Path::new(&path_str)
                .join(output_file_slug)
                .parent()
                .unwrap()
                .to_path_buf();

            // Use fresh Settings scope for each snapshot to avoid interference from prior or parallel test runs
            insta::with_settings!({
                snapshot_path => snapshot_path,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(output_file.file_name().unwrap().to_str(), output_content);
            });
        }

        let mut output_file_paths = output_files
            .iter()
            .map(|file| file.strip_prefix(output_dir).unwrap())
            .map(|file| file.to_string_lossy())
            .collect::<Vec<_>>();
        output_file_paths.sort();

        let snapshot_files: Vec<PathBuf> = WalkDir::new(snapshot_dir)
            .into_iter()
            .filter_map(|entry| entry.ok())
            .filter(|e| e.path().is_file())
            .filter(|e| e.file_name().to_str().unwrap().ends_with(".snap"))
            .map(|e| e.path().to_path_buf())
            .collect();

        let mut snapshot_file_paths = snapshot_files
            .iter()
            .map(|file| file.strip_prefix(snapshot_dir).unwrap())
            .map(|file| {
                file.to_string_lossy()
                    .strip_suffix(".snap")
                    .unwrap()
                    .to_string()
            })
            .collect::<Vec<_>>();
        snapshot_file_paths.sort();

        assert_eq!(snapshot_file_paths.join("\n"), output_file_paths.join("\n"));
    }

    fn read_file(file_path: &PathBuf) -> Vec<u8> {
        let mut file = fs::File::open(file_path).expect("Failed to open file");
        let mut content = Vec::new();
        file.read_to_end(&mut content).expect("Failed to read file");
        content
    }
}
