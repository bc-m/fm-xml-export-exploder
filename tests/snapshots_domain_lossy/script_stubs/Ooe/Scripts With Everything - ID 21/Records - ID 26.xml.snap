---
source: src/tests.rs
expression: output_content
---
<Script id="26" name="Records">
	<UUID modifications="2" userName="Mislav" accountName="admin" timestamp="2025-06-20T18:22:09">6C16E55C-DCC8-44D8-B79D-8D79AA9D4973</UUID>
	<Options hidden="False" access="ReadWrite" SiriShortcutVisible="False" runwithfullaccess="False" compatibility="0">0</Options>
</Script>
